optaplanner-examples/src/main/java/org/optaplanner/examples/tennis/optional/score/TennisConstraintProvider.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.tennis.optional.score;
+
+import org.optaplanner.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.Joiners;
+import org.optaplanner.examples.tennis.domain.TeamAssignment;
+import org.optaplanner.examples.tennis.domain.UnavailabilityPenalty;
+
+import static org.optaplanner.core.api.score.stream.ConstraintCollectors.loadBalanceByCount;
+
+public class TennisConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                oneAssignmentPerDatePerTeam(constraintFactory),
+                unavailabilityPenalty(constraintFactory)
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    private Constraint oneAssignmentPerDatePerTeam(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TeamAssignment.class)
+                .ifExistsOther(TeamAssignment.class,
+                        Joiners.equal(TeamAssignment::getTeam),
+                        Joiners.equal(TeamAssignment::getDay),
+                        Joiners.lessThan(TeamAssignment::getId),
+                        Joiners.filtering((teamAssignment, otherTeamAssignment) -> teamAssignment.getTeam() != null))
+                .penalize("oneAssignmentPerDatePerTeam", HardMediumSoftScore.ONE_HARD);
```

Comment:
This causes a "score trap" (see docs).

If 2 assignments conflict, the penality should be -1hard (AB)
If 3 assignements conflict, the penality should be -3hard (AB, BC, AC)
If 4 conflict, -6hard (AB, AC, AD, BC, BD, CD)

Reply:
Furtermore, the justification won't mention which 2 teamAssignments conflict.

Reply:
@ge0ffrey I'd argue this is outside the scope of the PR, which is to add CS for the existing DRL file (which has the "score trap" from what I can tell), but I agree with your point about how multiple assignment conflicts are penalized. I will add a TODO for a change of this rule (and its equivalent in the DRL file).

Reply:
The reasoning itself makes sense, but the DRL does it differently, the way I expected.

The DRL (= CS with join()) :

```
        TeamAssignment(team != null, $t : team, $d : day, $leftId : id)
        TeamAssignment(team == $t, day == $d, id > $leftId)
```

A DRL that matches the current PR (= CS with ifExists()) :

```
        TeamAssignment(team != null, $t : team, $d : day, $leftId : id)
        exists TeamAssignment(team == $t, day == $d, id > $leftId)
```

Reply:
assertionScoreDirectoryFactory might not always detect this difference.

Reply:
Ah, I see your point. Strange that a FULL_ASSERT didn't catch this, shouldn't the assertionScoreDirectoryFactory assert that the scores are penalized by the same values? In any case, I'll fix the CS.

Reply:
They're not penalized by the same value for a conflict of 3 entities or more. However, because it's hard constraint, on most datasets, it never going to evaluate a conflict of 3+ entities: as soon as it sees a conflict of 2 entities, it's very unlikely that that step will with the CH or LS, at least with the default set of move selectors.